src/github/github.installation.service.ts

```diff
@@ -0,0 +1,173 @@
+    const org = await this.allowedOrgRepository.findOne({
+      where: { organizationName, status: OrganizationStatus.ACTIVE },
+    });
+    return !!org;
+  }
+
+  async isUserAllowed(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<boolean> {
+    try {
+      const org = await this.findByName(organizationName);
+      if (org.status !== OrganizationStatus.ACTIVE) {
+        return false;
+      }
+      if (org.allowAllUsers) {
+        return true;
+      }
+      return org.activeUsers.includes(githubUsername);
+    } catch (error) {
+      this.logger.error(`Error checking user allowance: ${error.message}`);
+      return false;
+    }
+  }
+
+  async addAllowedUser(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    if (org.inactiveUsers.includes(githubUsername)) {
+      org.inactiveUsers = org.inactiveUsers.filter(
+        (username) => username !== githubUsername,
+      );
+    }
+    if (!org.activeUsers.includes(githubUsername)) {
+      org.activeUsers.push(githubUsername);
+      return this.allowedOrgRepository.save(org);
+    }
+    return org;
+  }
+
+  async removeAllowedUser(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    org.activeUsers = org.activeUsers.filter(
+      (username) => username !== githubUsername,
+    );
+    if (!org.inactiveUsers.includes(githubUsername)) {
+      org.inactiveUsers.push(githubUsername);
+    }
+    return this.allowedOrgRepository.save(org);
+  }
+
+  async updateStatus(
+    organizationName: string,
+    newStatus: OrganizationStatus,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    org.status = newStatus;
+    return this.allowedOrgRepository.save(org);
+  }
+
+  async uninstall(id: number) {
+    const now = Math.floor(Date.now() / 1000);
+
+    const pemFilename = Env.GITHUB.APP_PEM;
+    const pemFilePath = path.join(__dirname, '../../../', pemFilename);
+    const privateKey = fs.readFileSync(pemFilePath, 'utf8');
+
+    const payload = {
+      iat: now,
+      exp: now + 600,
+      iss: Env.GITHUB.APP_ID,
+      installation_id: id,
+    };
+    const token = jwt.sign(payload, privateKey, { algorithm: 'RS256' });
+    const octokit = new Octokit({
+      auth: token,
+    });
+
+    await octokit.request('DELETE /app/installations/{installation_id}', {
+      installation_id: id,
+      headers: {
+        'X-GitHub-Api-Version': '2022-11-28',
+      },
+    });
```

Comment:
PullProofAI

There is a potential performance issue in the `uninstall` method. The method generates a JWT token for each uninstallation request, which involves reading the private key file synchronously on each call. This synchronous file read operation can impact performance, especially under high load.

To improve performance, consider caching the private key in memory or loading it once during service initialization rather than reading it from disk every time `uninstall` is called.

```typescript
// Cache private key in memory during service initialization
private privateKey: string;

constructor() {
    const pemFilename = Env.GITHUB.APP_PEM;
    const pemFilePath = path.join(__dirname, '../../../', pemFilename);
    this.privateKey = fs.readFileSync(pemFilePath, 'utf8');
}

async uninstall(id: number) {
    const now = Math.floor(Date.now() / 1000);

    const payload = {
        iat: now,
        exp: now + 600,
        iss: Env.GITHUB.APP_ID,
        installation_id: id,
    };
    const token = jwt.sign(payload, this.privateKey, { algorithm: 'RS256' });
    const octokit = new Octokit({
        auth: token,
    });

    await octokit.request('DELETE /app/installations/{installation_id}', {
        installation_id: id,
        headers: {
            'X-GitHub-Api-Version': '2022-11-28',
        },
    });
}
```

By caching the private key, you reduce the overhead of reading the file on each call to `uninstall`, improving the overall performance of the service.


<!-- This is an auto-generated comment by PullProofAI -->